services/spar/src/Spar/Sem/SAML2/Library.hs

```diff
@@ -0,0 +1,137 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# OPTIONS_GHC -fplugin=Polysemy.Plugin #-}
+
+module Spar.Sem.SAML2.Library (saml2ToSaml2WebSso) where
+
+import qualified Control.Monad.Catch as Catch
+import Control.Monad.Except
+import Data.Id (TeamId)
+import Data.String.Conversions (cs)
+import Imports
+import Polysemy
+import Polysemy.Error
+import Polysemy.Input
+import Polysemy.Internal.Tactics
+import SAML2.WebSSO hiding (Error)
+import qualified SAML2.WebSSO as SAML hiding (Error)
+import qualified Spar.App as App
+import Spar.Error (SparCustomError (..), SparError)
+import Spar.Sem.AReqIDStore (AReqIDStore)
+import qualified Spar.Sem.AReqIDStore as AReqIDStore
+import Spar.Sem.AssIDStore (AssIDStore)
+import qualified Spar.Sem.AssIDStore as AssIDStore
+import qualified Spar.Sem.IdP as IdPEffect
+import Spar.Sem.Logger (Logger)
+import qualified Spar.Sem.Logger as Logger
+import Spar.Sem.SAML2
+import Wire.API.User.IdentityProvider (WireIdP)
+import Wire.API.User.Saml
+
+wrapMonadClientSPImpl :: Members '[Error SparError, Final IO] r => Sem r a -> SPImpl r a
+wrapMonadClientSPImpl action =
+  SPImpl $
+    action
+      `Catch.catch` (throw . SAML.CustomError . SparCassandraError . cs . show @SomeException)
+
+newtype SPImpl r a = SPImpl {unSPImpl :: Sem r a}
+  deriving (Functor, Applicative, Monad)
+
+instance Member (Input Opts) r => HasConfig (SPImpl r) where
+  getConfig = SPImpl $ inputs saml
+
+instance Members '[Input Opts, Logger String] r => HasLogger (SPImpl r) where
+  logger lvl = SPImpl . Logger.log lvl
+
+instance Member (Embed IO) r => MonadIO (SPImpl r) where
+  liftIO = SPImpl . embed @IO
+
+instance Member (Embed IO) r => HasCreateUUID (SPImpl r)
+
+instance Member (Embed IO) r => HasNow (SPImpl r)
+
+instance Members '[Error SparError, Final IO, AReqIDStore] r => SPStoreID AuthnRequest (SPImpl r) where
+  storeID = (wrapMonadClientSPImpl .) . AReqIDStore.store
+  unStoreID = wrapMonadClientSPImpl . AReqIDStore.unStore
+  isAliveID = wrapMonadClientSPImpl . AReqIDStore.isAlive
+
+instance Members '[Error SparError, Final IO, AssIDStore] r => SPStoreID Assertion (SPImpl r) where
+  storeID = (wrapMonadClientSPImpl .) . AssIDStore.store
+  unStoreID = wrapMonadClientSPImpl . AssIDStore.unStore
+  isAliveID = wrapMonadClientSPImpl . AssIDStore.isAlive
+
+instance Members '[Error SparError, IdPEffect.IdP, Final IO] r => SPStoreIdP SparError (SPImpl r) where
+  type IdPConfigExtra (SPImpl r) = WireIdP
+  type IdPConfigSPId (SPImpl r) = TeamId
+
+  storeIdPConfig = SPImpl . App.runSparInSem . App.storeIdPConfig
+  getIdPConfig = SPImpl . App.runSparInSem . App.getIdPConfig
+  getIdPConfigByIssuerOptionalSPId a = SPImpl . App.runSparInSem . App.getIdPConfigByIssuerOptionalSPId a
+
+instance Member (Error SparError) r => MonadError SparError (SPImpl r) where
+  throwError = SPImpl . throw
+  catchError m handler = SPImpl $ catch (unSPImpl m) $ unSPImpl . handler
+
+saml2ToSaml2WebSso ::
```

Comment:
I'm going to trust you on this, but:

```suggestion
-- | To learn more about polysemy tactics, read this:
--   * https://reasonablypolymorphic.com/blog/freer-higher-order-effects/
--   * https://reasonablypolymorphic.com/blog/tactics/
saml2ToSaml2WebSso ::
```

Reply:
It's nightmarish code, sorry. Polysemy v2 promises to clean this up, and in my offtime I'm going to port this function to see how improved things are.

Happy to write some comments about WTF is going on here.

Reply:
great idea.  if you're doing this right, perhaps you can point other people to this when they're asking about tactics?  :)